servicetalk-encoding-netty/src/main/java/io/servicetalk/encoding/netty/NettyChannelContentCodec.java

```diff
@@ -0,0 +1,335 @@
+
+                    private final MessageToByteEncoder<ByteBuf> encoder = encoderSupplier.get();
+                    private final EmbeddedChannel channel = newEmbeddedChannel(encoder, allocator);
+
+                    @Override
+                    public void onSubscribe(PublisherSource.Subscription subscription) {
+                        subscriber.onSubscribe(subscription);
+                    }
+
+                    @Override
+                    public void onNext(Buffer next) {
+                        if (next == null) {
+                            return;
+                        }
+
+                        // onNext will produce AT-MOST N items (from upstream)
+                        // +1 for the encoding footer (ie. END_OF_STREAM)
+                        if (next == END_OF_STREAM) {
+                            // May produce footer
+                            preparePendingData(channel);
+                            Buffer buffer = drainChannelOutToSingleBuffer(channel, allocator);
+                            if (buffer != null) {
+                                subscriber.onNext(buffer);
+                            }
+
+                            return;
+                        }
+
+                        channel.writeOutbound(toByteBuf(next));
+                        Buffer buffer = drainChannelOutToSingleBuffer(channel, allocator);
+                        if (buffer != null) {
+                            subscriber.onNext(buffer);
+                        }
+                    }
+
+                    @Override
+                    public void onError(Throwable t) {
+                        safeCleanup(channel);
+                        subscriber.onError(t);
+                    }
+
+                    @Override
+                    public void onComplete() {
+                        cleanup(channel);
+                        subscriber.onComplete();
+                    }
+                });
+    }
+
+    @Override
+    public Buffer decode(final Buffer src, final int offset, final int length, final BufferAllocator allocator) {
+        requireNonNull(src);
+        final int availableBytes = src.readableBytes() - offset;
+        if (offset < 0) {
+            throw new IllegalArgumentException("Invalid offset: " + offset + " (expected >= 0)");
+        }
+
+        if (length <= 0) {
+            throw new IllegalArgumentException("Invalid length: " + length + " (expected > 0)");
+        }
+
+        if (length > availableBytes) {
+            throw new IllegalStateException("Invalid length: " + length + " (expected <= " + availableBytes);
+        }
+
+        final ByteToMessageDecoder decoder = decoderSupplier.get();
+        final EmbeddedChannel channel = newEmbeddedChannel(decoder, allocator);
+
+        try {
+            ByteBuf origin = toByteBuf(src, offset, length);
+            channel.writeInbound(origin);
+            return newBufferFrom(channel.readInbound());
```

Comment:
It's possible that netty's decoder will produce multiple output buffers, we need to read all of them